### PR TITLE
Include detailed tests for forcefield metadata

### PR DESCRIPTION
**Pull request by justinGilmer**
### PR Summary:
Previously, the tests in test_forcefield.py were assumed to test that
forcefield metadata like `version`, `name`, etc. were properly loaded
into the `Forcefield` object.

This is only partially true, as there are cases where the files either
go out of scope before that information is gathered, etc. This PR
includes additional tests to ensure that this information is not being
lost prematurely. Big thanks to: @mattwthompson, @ahy3nz for discovering
this issue, creating MWE's for some of these cases, and taking my
investigation further and most likely pinpointing the issue.

### PR Checklist
------------
 - [x] Includes appropriate unit test(s)
 - [ ] Appropriate docstring(s) are added/updated
 - [x] Code is (approximately) PEP8 compliant
 - [x] Issue(s) raised/addressed?

Refer to issues/PRs: #433 #434 

Note: This does not have @ahy3nz 's fix incorporated yet, just the test cases that fail and currently work.

### CODE REVIEW

**Comment by mattwthompson** [comment]
Looks like this reproduces the issue (`test_load_metadata_from_internal_name`, the one using `Forcefield(name="")` is failing).

I took Alex's suggested patch (https://github.com/mosdef-hub/foyer/issues/434#issuecomment-870754409) and made a couple fixes so that tests would pass. I haven't thought much about unintended side effects yet.

Here is the commit; feel free to move it around, copy, etc. : https://github.com/mosdef-hub/foyer/commit/22855d422f5770f596356e3cb0d48b49a899f3b9

**Comment by justinGilmer** [comment]
Yeah we essentially had the same design process and handling of the strings. Once this passes and we have the approvals, we can merge. Thanks @mattwthompson  and @ahy3nz for both of your contributions!

**Comment by mattwthompson** [comment]
Does it matter that you're using `preprocessed_files` and I'm using `all_files_to_load`? Both seem to pass tests ... it's probably better to use the pre-processed ones in principle?

**Comment by mattwthompson** [comment]
(Sorry for the misclick 😬)

**Comment by justinGilmer** [comment]
> Does it matter that you're using `preprocessed_files` and I'm using `all_files_to_load`? Both seem to pass tests ... it's probably better to use the pre-processed ones in principle?

That was my thought, we already strip out/handle things that might not behave the way we expect once we have pre-processed.

**Review by mattwthompson — APPROVED** [review]
LGTM iff tests pass.

There should be some added confidence if another piece of metadata can be added (#433) without re-engineering this machinery.

**Comment by ahy3nz** [comment]
>>Does it matter that you're using preprocessed_files and I'm using all_files_to_load? Both seem to pass tests ... it's probably better to use the pre-processed ones in principle?

> That was my thought, we already strip out/handle things that might not behave the way we expect once we have pre-processed.

While I agree the pre-processed XML files are safer to try reading, we might have to consider this block of code that is being deleted in this PR. Originally, IIRC @justinGilmer ran into some problems with `/tmp` blowing up when running a lot of atomtyping, so we needed to add this code chunk to remove old XML files. If this chunk of code is removed, this problem may come up again.

https://github.com/mosdef-hub/foyer/commit/748bdd8646342e7a253ee3d3c9944eaad6b83cc9
```python3
        try:
            super(Forcefield, self).__init__(*preprocessed_files)
        finally:
            for ff_file_name in preprocessed_files:
                os.remove(ff_file_name)
```

**Comment by justinGilmer** [comment]
I did add a small section at the end of this init method to iterate through the open files and os.remove them. It runs into the problem where I cant really put a try catch finally around the entire part of that code to always ensure that the files are deleted, but it should handle most cases.

https://github.com/mosdef-hub/foyer/pull/435/files#diff-2e32e78335ccab485367397b205f395d5793e27101a8e63226d79081eb4f2356R536-R537

**Comment by ahy3nz** [comment]
My mistake! Good catch

**Review by ahy3nz — APPROVED** [review]

**Review by daico007 — APPROVED** [review]
LGTM!

**Pull request merged**
